storage/core/src/types.rs

```diff
@@ -0,0 +1,185 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Database backend support building blocks
+
+use core::ops::Range;
+
+/// Used to identify particular key-value map in the database
+#[derive(Eq, PartialEq, PartialOrd, Ord, Clone, Copy, Debug)]
+pub struct DbMapId(usize);
+
+impl DbMapId {
+    /// New index
+    pub const fn new(id: usize) -> Self {
+        DbMapId(id)
+    }
+
+    /// Get the index as usize
+    pub const fn as_usize(&self) -> usize {
+        self.0
+    }
+}
+
+/// Number of key-value maps in the database
+#[derive(Eq, PartialEq, PartialOrd, Ord, Clone, Copy, Debug)]
+pub struct DbMapCount(usize);
+
+impl DbMapCount {
+    /// Convert to `usize`
+    pub const fn as_usize(self) -> usize {
+        self.0
+    }
+
+    /// Iterator over all database indices
+    pub fn indices(self) -> impl Iterator<Item = DbMapId> + ExactSizeIterator {
+        (0..self.0).map(DbMapId::new)
+    }
+}
+
+/// Associate data of type `T` with each DB map
+///
+/// This is used to associate certain data with each DB map. The data can be metadata containing
+/// names and other information about the maps, like [DbMapDesc]. It can also be used by the
+/// backend to store for example handles to open key-value maps.
+///
+/// The interface of this type is deliberately constrained to make it hard to use inappropriately.
+/// It can be constructed in two ways:
+/// 1. The [Self::new] method takes in the number of DB maps. The map count is in a type wrapper to
+///    make it harder to use arbitrary number there.
+/// 2. The [Self::transform] method is used to derive one piece of data from another. It preserves
+///    the data count and ensures the DB map IDs remain consistent between the original and the
+///    derived [DbMapsData] instance.
+/// 3. The data can only be accessed by a [DbMapId] so it's only indexed into using a valid DB
+///    map identifier.
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub struct DbMapsData<T>(Vec<T>);
+
+impl<T> DbMapsData<T> {
+    /// New maps data for given number of key-value maps and initialization function
+    pub fn new(map_count: DbMapCount, constructor: impl FnMut(DbMapId) -> T) -> Self {
+        Self(map_count.indices().map(constructor).collect())
+    }
+
+    /// Get the number of maps
+    pub fn db_map_count(&self) -> DbMapCount {
+        DbMapCount(self.0.len())
+    }
+
+    /// Apply given function to each map and collect the results
+    pub fn transform<U>(&self, func: impl FnMut(&T) -> U) -> DbMapsData<U> {
```

Comment:
Isn't it easier to just implement IntoIterator for this type and then get all these functions for free?

Reply:
I considered that but in the end decided for a more limited interface. Iterator is allowed to reorder, drop, duplicate elements, all sorts of stuff like that. This ensures the map IDs match between the original `DbMapData` and the resulting one.